src/librustc/ty/fast_reject.rs

```diff
@@ -101,3 +102,80 @@ pub fn simplify_type<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>,
         ty::TyInfer(_) | ty::TyError => None,
     }
 }
+
+// A version of fast_reject::SimplifiedType that allows for stable hashing and
+// especially can act as a stable sorting key.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Clone)]
+pub enum StableSimplifiedType {
```

Comment:
I don't quite understand why we need two versions here -- can we just use this version? Glancing around the code, I didn't see any obvious reason why not.

Reply:
Right. Changing this would make constructing simplified types slightly more expensive because of having to look up the `DefPathHash`. I'm not sure how I feel about that.